app.py

```diff
@@ -28,18 +24,72 @@ def inject_func():
                 len=len)
 
 
-## initiate ###################
+###### before and after each request ########
+@app.before_request
+def open_conn():
+    global connections
+    conn = connections.getconn()
+    cursor = conn.cursor(cursor_factory = DictCursor)
+    g.cur = cursor
+    g.conn = conn
+    g.error = False
+
+
+@app.after_request
+def close_conn(response):
+    
+    if g.conn is not None:    
```

Comment:
Do we have a case where `g.conn` is None?

Reply:
Yes when we are redirected to error handler view

Reply:
In that case, are you saying the `before_request` hook is not executed?

Reply:
I am confused here!
when we get an error and we are redirected to errorhandler, it functions as a view function and our after request shouldn't run but it runs and I get the error that conn is None. that's why I put a logic at the beginning of the after_request . Check this out: https://pythonise.com/series/learning-flask/python-before-after-request


after_request functions will be run after every request and provide access to the request context, meaning we still have access to ssession and g. For example:

@app.after_request
def after_request_func(response):
    username = g.username
    foo = session.get("foo")
    print("after_request is running!", username, foo)
    return response
At this point you might be thinking; This is a great place to do something like close a database connection, and you'd be right.

**However something to note is that any functions decorated with after_request will NOT run if your application throws an exception.**


I think this guy is wrong! my after_request is running even after exceptions